app-backend/api/src/main/scala/uploads/Routes.scala

```diff
@@ -51,240 +56,292 @@ trait UploadRoutes
+          val potentialNewBytes =
+            newUpload.fileSizeBytes.getOrElse(0.toLong).max(bytesUploaded)
+          val uploadToInsert =
+            (
+              newUpload.uploadType,
+              newUpload.source,
+              newUpload.fileType
+            ) match {
+              case (UploadType.S3, _, FileType.NonSpatial) => {
+                if (newUpload.files.nonEmpty) newUpload
                 else
                   throw new IllegalStateException(
-                    "No acceptable files found in the provided source"
+                    "S3 upload must specify a source if no files are specified"
                   )
               }
-            }
-            case (UploadType.S3, None, _) => {
-              if (newUpload.files.nonEmpty) newUpload
-              else
+              case (UploadType.S3, Some(source), _) => {
+                if (newUpload.files.nonEmpty) newUpload
+                else {
+                  val files = listAllowedFilesInS3Source(source)
+                  if (files.nonEmpty) newUpload.copy(files = files)
+                  else
+                    throw new IllegalStateException(
+                      "No acceptable files found in the provided source"
+                    )
+                }
+              }
+              case (UploadType.S3, None, _) => {
+                if (newUpload.files.nonEmpty) newUpload
+                else
+                  throw new IllegalStateException(
+                    "S3 upload must specify a source if no files are specified"
+                  )
+              }
+              case (_, _, FileType.GeoJson) => {
                 throw new IllegalStateException(
-                  "S3 upload must specify a source if no files are specified"
+                  "Scene uploads must contain imagery, not GeoJson"
                 )
+              }
+              case (_, _, _) => {
+                if (newUpload.files.nonEmpty) newUpload
+                else
+                  throw new IllegalStateException(
+                    "Remote repository upload must specify some ids or files"
+                  )
+              }
             }
-            case (_, _, FileType.GeoJson) => {
-              throw new IllegalStateException(
-                "Scene uploads must contain imagery, not GeoJson"
-              )
-            }
-            case (_, _, _) => {
-              if (newUpload.files.nonEmpty) newUpload
-              else
-                throw new IllegalStateException(
-                  "Remote repository upload must specify some ids or files"
-                )
+          authorizeScopeLimit(
+            Future.successful(potentialNewBytes),
+            Domain.Uploads,
+            Action.Create,
+            user
+          ) {
+            onComplete {
+              for {
+                uploadOpt <- UploadDao
+                  .insert(uploadToInsert, user, potentialNewBytes)
+                  .map(Some(_))
```

Comment:
I think you don't need this to be an option anymore -- the optionality was only present to handle `isBelow`, which is now covered in the authorization check